_published/0017-azuredevops-repos-and-not-projects.md

```diff
@@ -0,0 +1,259 @@
+response for the [repository endpoints](https://docs.microsoft.com/en-us/rest/api/azure/devops/git/repositories?view=azure-devops-server-rest-5.0).
+
+This behavior exists in wharf-provider-azuredevops in v1. Starting from v2,
+this behavior is as explained below in this RFC.
+
+### Import repositories, not projects
+
+Each code repository is imported 1:1 into Wharf as distinct Wharf projects.
+If an Azure DevOps project contains more than 1 code repository, then said
+Azure DevOps project will in practice be imported the same number of times,
+albeit with different names on the Wharf projects to differentiate between them.
+
+### Wharf project naming
+
+When importing an Azure DevOps project, the Wharf group and project names use
+the following format:
+
+| № code repos  | Wharf project group           | Wharf project name |
+| ------------  | -------------------           | ------------------ |
+| 0             | *not imported*                | *not imported*     |
+| 1             | `{azure-org}`                 | `{azure-project}`  |
+| 2 *(or more)* | `{azure-org}/{azure-project}` | `{azure-repo}`     |
+
+Where:
+
+- `{azure-org}` = name of the Azure DevOps organization that holds the Azure
+  DevOps project.
+
+- `{azure-project}` = name of the Azure DevOps project.
+
+- `{azure-repo}` = name of the code repository inside the Azure DevOps project.
+
+Wharf projects that has been imported in earlier versions may need an automatic
+rename on their next import. To see how this is handled in detail, see the
+[#Renaming projects](#renaming-projects) section.
+
+### Import procedure
+
+1. First get the Azure DevOps project(s) using the endpoint:
+
+   ```http
+   GET https://dev.azure.com/{organization}/_apis/projects
+   ```
+
+   <https://docs.microsoft.com/en-us/rest/api/azure/devops/core/projects/list?view=azure-devops-rest-5.0>
+
+   This is used to get the following information:
+
+   - Project name
+   - Project description
+   - Project ID (GUID)
+
+2. For every project, list the code repositories to import using the endpoint:
+
+   ```http
+   GET https://dev.azure.com/{organization}/{projectId}/_apis/git/repositories
+   ```
+
+   <https://docs.microsoft.com/en-us/rest/api/azure/devops/git/repositories/list?view=azure-devops-rest-5.0>
+
+   This is used to get the following information:
+
+   - Repository name
+   - Git SSH URL
+   - Default branch name (but make sure to trim away the `refs/heads/` prefix)
+
+3. For every code repository, list the branches to import using the endpoint:
+
+   ```http
+   GET https://dev.azure.com/{organization}/_apis/git/repositories/{repositoryId}/refs?filter=heads/
+   ```
+
+   <https://docs.microsoft.com/en-us/rest/api/azure/devops/git/refs/list#refs-heads>
+
+   This is used to get the following information:
+
+   - Branch name (but make sure to trim away the `refs/heads/` prefix)
+
+## Compatibility
```

Comment:
This deserves a note. My gut feeling was that there would be two different types of importers for local instances or private cloud and public cloud instances of azure dev ops. 

Does this functionality currently work on private cloud instances? Will it continue working on private cloud instances after this change? Does this RFC convert a private cloud provider into a public cloud provider because the two are incompatible?

Reply:
Excellent question! I'll add a well deserved note :)

Reply:
Note added. At the time of writing, this is what I came up with:

### Keeping compatibility with self-hosted instances

Will be kept. Compared to v1, we will use 1 new endpoint:

```http
GET https://dev.azure.com/{organization}/{projectId}/_apis/git/repositories
```

This new endpoint does also exist in Azure DevOps API v5.0, which we have
relied on up until now.

As for the new procedure of getting the Git SSH URL, this will produce better
stability than before for unconventional installations of self-hosted
Azure DevOps instances as long as their Git SSH URL is configured correctly,
whereas conventional installations will continue to behave as before.